src/robust_provider.rs

```diff
@@ -172,23 +213,76 @@ impl<N: Network> RobustProvider<N> {
     /// the entire operation (including time spent inside the RPC call) cannot exceed
     /// `max_timeout`.
     ///
+    /// If the timeout is exceeded and fallback providers are available, it will
+    /// attempt to use each fallback provider in sequence.
+    ///
     /// # Errors
     ///
-    /// - Returns [`RpcError<TransportErrorKind>`] with message "total operation timeout exceeded"
-    ///   if the overall timeout elapses.
+    /// - Returns [`RpcError<TransportErrorKind>`] with message "total operation timeout exceeded
+    ///   and all fallback providers failed" if the overall timeout elapses and no fallback
+    ///   providers succeed.
     /// - Propagates any [`RpcError<TransportErrorKind>`] from the underlying retries.
     async fn retry_with_total_timeout<T, F, Fut>(&self, operation: F) -> Result<T, Error>
     where
-        F: Fn() -> Fut,
+        F: Fn(RootProvider<N>) -> Fut,
+        Fut: Future<Output = Result<T, RpcError<TransportErrorKind>>>,
+    {
+        // Try primary provider first
+        let result = self.try_provider_with_timeout(&self.provider, &operation).await;
+
+        if let Ok(value) = result {
+            return Ok(value);
+        }
+
+        if result.is_err() && self.fallback_providers.is_empty() {
```

Comment:
```suggestion
        if self.fallback_providers.is_empty() {
```
nit: if the execution gets here, `result` [must be an error](https://github.com/OpenZeppelin/Event-Scanner/pull/140/files#diff-7c3abb68ead49d79c137d9cc6fde16d55052445640ed7eaaec0411546eea5276R233-R237)

Reply:
true good catch